src/util.rs

```diff
@@ -35,17 +35,18 @@ impl<T> MMRStoreWriteOps<T> for &MemStore<T> {
 
 pub type MemMMR<'a, T, M> = MMR<T, M, &'a MemStore<T>>;
 
-pub trait VeqDequeExt<T: Ord> {
-    fn insert_sorted(&mut self, value: T);
+pub trait BTreeMapExt<K, V> {
+    fn checked_insert(&mut self, key: K, value: V) -> bool;
 }
 
-impl<T: Ord> VeqDequeExt<T> for VecDeque<T> {
-    fn insert_sorted(&mut self, value: T) {
-        match self.binary_search(&value) {
-            Ok(_pos) => {
-                // element already in vector @ `pos`
-            }
-            Err(pos) => self.insert(pos, value),
+impl<K: Ord, V: PartialEq> BTreeMapExt<K, V> for BTreeMap<K, V> {
+    fn checked_insert(&mut self, key: K, value: V) -> bool {
+        let old_value = self.get(&key);
+        if old_value != None && old_value != Some(&value) {
+            return false;
         }
+
+        self.insert(key, value);
+        true
```

Comment:
I'd suggest using entry API: https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.entry to avoid doing a lookup before insertion (or alternatively, clone the value and do a single insert)

Reply:
Done